Add grounded object node

vsbogd opened the pull request:
Implementation of design discussed in issue https://github.com/singnet/atomspace/issues/43 
`GroundedObjectNode`, `GetMethodLink` and `SnetExecutionOutputLink` are added to represent object in the atomspace atoms.

Things done:
- `GroundedObject` interface is added, all grounded objects should implement this interface to be represented in the atomspace
- `GroundedFunctionLink` is added to represent the function which can be executed by `SnetExecutionOutputLink`
- `GetMethodLink` is added as implementation of `GroundedFunctionLink` which represents method of `GroundedObjectNode`, method name is provided by `ConceptNode` 
- `SnetExecutionoutputLink` is added to execute `GroundedFunctionLink`, `SnetExecutionOutputLink` is backward compatible with `ExecutionOutputLink` and can be used instead of it
- Python wrapper for `GroundedObjectNode` is added to represent Python objects inside of OpenCog

To be done:
- documentation
- more unit tests

Any suggestions on better naming are welcome, in particular:
- better name for `SnetExecutionOutptuLink` (may be `ExecutionOutputExLink` ?)
- better name for `GetMethodLink` (former `DotLink` may be `GroundedMethodLink` ?)

vsbogd commented:
@noskill, @Necr0X0der please take a look.

ngeiswei commented on opencog/atoms/execution/GetMethodLink.cc:
```diff
@@ -0,0 +1,60 @@
+	: GroundedFunctionLink(output_set, type)
+{
+	check_outgoing_type(0, GROUNDED_OBJECT_NODE);
+	check_outgoing_type(1, CONCEPT_NODE);
```
Why are you checking here as opposed to a validator (I don't have an opinion on the matter, just curious)?

vsbogd replied:
Actually it is a question for me which way to use. Most of atomspace code uses checks in constructors, like recent `CondLink`. And validators are used for three types only.

BTW it is a place where I have enchanced atomspace outside of opencog repository, I mean I have added `OrderedLink` class to implement arguments check.

vsbogd replied:
Not sure `OrderedLink` is needed actually, will look at alternative validator implementation to not introduce more inconsistencies then it is needed.

vsbogd replied:
Actually there is not standard way to define Validator near the Atom which uses the validator, you need to add it into Checkers.cc which is ugly. So I am going to make a fix in the opencog repo. And I just will move all of the code into GetMethodLink.cc before that.

ngeiswei replied:
I agree that it's ugly. If you know how to beautify it, it's welcome :-)

Necr0x0Der commented:
* `GetMethodLink` - sounds OK to me. Possible alternatives: `AccessorLink`, `ObjectFieldLink`, `(Object)AttributeLink`, `(Object)GetterLink`, etc. don't sound better to me. I only don't like `DotLink` too much, because "dot" is too polysemantic.
* `SnetExecutionOutputLink` - well, it's OK as a temporary name, because it makes clear that this is a replacement/extension of `ExecutionOutputLink` for stuff that works in snet, but not original repo. However, in general, I don't like both "ExecutionOutput" name and "Snet" prefix, so I would propose to fix these both at once. Since `GetMethodLink` is an implementation of `GroundedFunctionLink`, and `SnetExecutionOutputLink` can become a general method for execution functions, I'd call it `ApplicationLink` (I believe Atomese needs a generic `ApplicationLink`, but not sure if `SnetExecutionOutputLink` is or can become such a link); one might also want to call it `FunctionApplicationLink` or `GroundedApplicationLink`, but it's a pain to write these long names. If @ngeiswei doesn't agree that it can be called `ApplicationLink`, we can think about another naming.

I'd also say that requiring the second argument to be a concept node (`check_outgoing_type(1, CONCEPT_NODE);`) is too restrictive. We can actually accept any Atom with name, i.e. any Node. What do you think? I mean I don't see any reason why the method name should be a concept. One might argue that it should be `DefinedSchemaNode` or whatever for a stronger type restriction during pattern matching (if we'd like to enumerate all methods).

Here are some additional thoughts (not for this PR, but just not to forget):
* It would be nice to add (a lot of) examples how `GroundedObjectNode` can be useful (e.g. showing how to use DNNs from python in Scheme, accessing data structures in a semi-declarative way - accessing object fields instead of methods since they are accessed through getter methods, etc.);
* We would like to have a smart and flexible caching mechanism for calculated values. Right now, we have to implement an awkward caching mechanism on the "client" side with temporary atomspaces, which are populated with temporary atomes, to which cached values are attached (e.g., `atomspace.add_node(types.ConceptNode, 'tmp-result-' + str(id))`... [here](https://github.com/singnet/semantic-vision/blob/master/experiments/opencog/cog_module/module.py)). I dreamed about such a mechanism since [this issue](https://github.com/opencog/atomspace/issues/1868) or even earlier (don't remember where I described the problem for the first time and proposed to attach values to execution output links in do_execute, which was rejected as a non-issue). Now we have a chance to solve this problem in a non-hacky way...

vsbogd commented:
> `ApplicationLink`

Sounds good to me. @ngeiswei , what do you think?

> I'd also say that requiring the second argument to be a concept node (`check_outgoing_type(1, CONCEPT_NODE);`) is too restrictive.

Yes, I thought about it while writing code but I supposed that adding any atom may be less strict and clear. May be it is worth to add new node to designate methods and functions, like `FunctionNameNode` just for type checking.

>  It would be nice to add (a lot of) examples how `GroundedObjectNode` can be useful

Yes, I am going to add the examples and documentation in separate PR.

> We would like to have a smart and flexible caching mechanism for calculated values.

I will look at semantic-vision code closer and will return with my suggestions regarding caching design.

Necr0x0Der commented:
> Yes, I thought about it while writing code but I supposed that adding any atom may be less strict and clear. May be it is worth to add new node to designate methods and functions, like `FunctionNameNode` just for type checking.

Well, I don't like adding more types like `FunctionNameNode`. Indeed, it would be easier for PM to enumerate different method names with this type, but users can simply say that some concept nodes corresponding to function names inherit from `(ConceptNode "Function")` or even from `(ConceptNode "MyObject methods")` (although I'd prefer to have a way to automatically derive this inheritance without the necessity for the user to write it down in his code, but this is another question). My point is that it can be up to the user, what type to use here, because what we need is just a string name. Maybe the user will want to use Nodes with method names in some other places, which will require some other specific types, and this will simply don't work... So, it's better to accept as general type as possible, imo.

ngeiswei commented:
- I'm perfectly OK with `SnetExecutionOutputLink` -> `ApplicationLink`!!! I love it even. I guess it could also be `ApplyLink` or such, I'm neutral.
- Regarding the method name atom type, I would think it should either be any atom, or a specialized type. It doesn't make much sense to be just a `ConceptNode` because a concept node has the property of being a set, with members attached to it via `MemberLink`, and I don't see where that property plays a role here.

Necr0x0Der commented:
> I guess it could also be `ApplyLink` or such
`ApplyLink` is even more concise. Sounds good to me.

ngeiswei commented:
I suppose `MethodLink` could be possible as well.

vsbogd commented:
I have renamed `SnetExecutionOutputLink` to the `ApplyLink`. Allowed any `Node` to contain the method name. And added a way to construct empty GroundedObjectNode as we discussed previously with @Necr0x0Der. I would merge it and add unit tests and documentation in further PRs.

Necr0x0Der commented:
> I suppose `MethodLink` could be possible as well.

```
Apply
    Method[Of]
        GroundedObject "some-object"
        MethodName "some-method"
     List args...
```
looks quite readable. So, I'm ok with renaming GetMethodLink to just MethodLink (although without Link, MethodOf sounds more understandable). However, GetMethod is also OK to me. Not sure if some variant is better...

>  Regarding the method name atom type, I would think it should either be any atom, or a specialized type.

With a specialized type, it becomes more readable indeed, but I still don't like it for other reasons...

noskill commented on opencog/cython/opencog/grounded_object_node.pyx:
```diff
@@ -0,0 +1,48 @@
+    def set_object(self, obj):
+        cdef shared_ptr[cGroundedObject] o_ptr
+        o_ptr.reset(new cPythonGroundedObject(<PyObject*>obj))
+        self.get_c_grounded_object_node_ptr().set_object(o_ptr)
```
I think deference would work here instead of raw pointer:
 
deref(self.get_c_value_ptr()).set_object(o_ptr)

noskill commented:
Look good!
